libmamba/include/mamba/core/package_fetcher.hpp

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2023, QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+#ifndef MAMBA_CORE_PACKAGE_FETCHER_HPP
+#define MAMBA_CORE_PACKAGE_FETCHER_HPP
+
+#include <functional>
+
+#include "mamba/core/channel.hpp"
+#include "mamba/core/download.hpp"
+#include "mamba/core/package_cache.hpp"
+#include "mamba/core/package_handling.hpp"
+#include "mamba/core/package_info.hpp"
+#include "mamba/core/thread_utils.hpp"
+
+namespace mamba
+{
+
+    class PackageFetcher;
+
+    enum class PackageExtractEvent
+    {
+        validate_update,
+        validate_success,
+        validate_failure,
+        extract_update,
+        extract_success,
+        extract_failure
+    };
+
+    class PackageExtractRequest
```

Comment:
Yeah I agree the name is confusing, I will change that.

Reply:
To me "task" only means something that runs like a fonction, might be embarking it's data to work on. It's an element processed by an scheduling and/or asynchronous or synchronous system, like a value is to a container. So it would work well in that context.
In general I dont like "runner" for tasks, it's ambiguous with the role of a scheduler and/or executor and/or execution context.

Other alternatives: operation, "PackageExtracter" (as in it's role is to do the extract)... 
You decide 👍🏽

Reply:
OK, then I'm probably biased on this ;) Let's use PackageExtractTask then.